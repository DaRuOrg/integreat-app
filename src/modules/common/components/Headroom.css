@import url('../../common/constants/colors.css');

.headroom {
  position: sticky;
  top: 0;
  z-index: 1;
  will-change: transform;
}
