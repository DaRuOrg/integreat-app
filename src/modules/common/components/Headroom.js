import React from 'react'
import PropTypes from 'prop-types'
import raf from 'raf'

import style from './Headroom.css'

class Headroom extends React.PureComponent {
  static propTypes = {
    children: PropTypes.any.isRequired,
    scrollHeight: PropTypes.number.isRequired,
    pinStart: PropTypes.number.isRequired,
    stickyAncestor: PropTypes.node,
    height: PropTypes.number
  }

  static defaultProps = {
    pinStart: 0
  }

  constructor (props) {
    super(props)
    this.handleEvent = this.handleEvent.bind(this)
    this.update = this.update.bind(this)
    this.ticking = false
    this.lastScrollY = 0
    this.lastTransform = 0
    this.lastKnownScrollY = 0
    this.currentScrollY = 0
    this.direction = 'down'
    this.state = {transform: 0}
  }

  componentDidMount () {
    window.addEventListener('scroll', this.handleEvent)
  }

  componentWillUnmount () {
    window.removeEventListener('scroll touchstart touchmove', this.handleEvent)
  }

  getScrollY = () => {
    if (window.pageYOffset !== undefined) {
      return window.pageYOffset
    } else if (window.scrollTop !== undefined) {
      return window.scrollTop
    } else {
      return (document.documentElement || document.body.parentNode || document.body).scrollTop
    }
  }

  truncate (transform, currentScrollY) {
    if (transform + currentScrollY < this.props.pinStart) {
      return 0
    }
    return Math.min(0, Math.max(-this.props.scrollHeight, transform))
  }

  update () {
    // this.lastScrollY: ScrollTop when the scroll direction last changed
    // this.lastTransform: Transform when the scroll direction last changed
    // this.lastKnownScrollY: Last ScrollTop (to detect change in scroll direction)
    // this.currentScrollY: Current ScrollTop
    // this.direction: 'up' || 'down'

    this.currentScrollY = this.getScrollY()
    if (this.lastKnownScrollY < this.currentScrollY) { // We're moving up
      if (this.direction !== 'up') { // We've changed direction from down to up!
        this.lastTransform = this.state.transform
        this.lastScrollY = this.lastKnownScrollY
        this.direction = 'down'
      }
    } else if (this.lastKnownScrollY > this.currentScrollY) { // We're moving down
      if (this.direction !== 'down') { // We've changed direction from up to down!
        this.lastTransform = this.state.transform
        this.lastScrollY = this.lastKnownScrollY
        this.direction = 'up'
      }
    }
    this.lastKnownScrollY = this.currentScrollY
    const newTransform = this.truncate(this.lastTransform + (this.lastScrollY - this.currentScrollY), this.currentScrollY)
    if (newTransform !== this.state.transform) {
      this.setState({transform: newTransform})
    }
    this.ticking = false
  }

  handleEvent () {
    if (!this.ticking) {
      this.ticking = true
      raf(this.update)
    }
  }

  render () {
    const {height, stickyAncestor, children} = this.props
    const {transform} = this.state
    return <React.Fragment>
      <div className={style.headroom2}
                  style={{transform: `translateY(${transform}px)`}}>
        {children}
      </div>
      {stickyAncestor && React.cloneElement(stickyAncestor, {stickyTop: height + transform})}
    </React.Fragment>
  }
}

export default Headroom
