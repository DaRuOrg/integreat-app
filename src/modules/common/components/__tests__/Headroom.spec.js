import React from 'react'
import { mount, shallow } from 'enzyme'
import Headroom from '../Headroom'
import raf from 'raf'

jest.mock('raf', () => jest.fn())

describe('Headroom', () => {
  const MockNode = () => <div />
  const MockAncestor = () => <div />

  const createComponent = props => mount(
    <Headroom scrollHeight={50}
              height={100}
              stickyAncestor={<MockAncestor />}
              {...props}>
      <MockNode />
    </Headroom>)

  test('should have correct default state', () => {
    const component = createComponent()
    expect(component.state()).toEqual({transform: 0, stickyTop: 0})
    expect(component.prop('pinStart')).toEqual(0)
  })

  test('should render with values from its state', () => {
    const component = createComponent()
    component.setState({transform: 42, stickyTop: 24})
    expect(component).toMatchSnapshot()
  })

  test('should render with no stickyAncestor supplied', () => {
    const component = shallow(
      <Headroom scrollHeight={50}><MockNode /></Headroom>)
    expect(component).toMatchSnapshot()
  })

  test('should attach and detach listener for onscroll event', () => {
    const originalAdd = window.addEventListener
    const originalRemove = window.removeEventListener
    window.addEventListener = jest.fn()
    window.removeEventListener = jest.fn()

    const component = mount(<Headroom scrollHeight={50}><MockNode /></Headroom>)
    const handleEventCallback = component.instance().handleEvent
    expect(window.addEventListener).toHaveBeenCalledWith('scroll', handleEventCallback)
    component.unmount()
    expect(window.removeEventListener).toHaveBeenCalledWith('scroll', handleEventCallback)

    window.addEventListener = originalAdd
    window.removeEventListener = originalRemove
  })

  test('should lock, request animation frame and update, on handleEvent', () => {
    const component = mount(<Headroom scrollHeight={50}><MockNode /></Headroom>)
    component.instance().update = jest.fn()

    // Call first time
    component.instance().handleEvent()
    expect(raf).toHaveBeenCalledTimes(1)
    // Raf has not yet been performed
    expect(component.instance().update).not.toHaveBeenCalled()

    // Check that we cannot call Raf a second time (as long as it hasn't been performed)
    component.instance().handleEvent()
    expect(raf.mock.calls).toHaveLength(1)

    // Now perform the raf
    raf.mock.calls[0][0]()
    expect(component.instance().update).toHaveBeenCalledTimes(1)

    // Should have unlocked again
    component.instance().handleEvent()
    expect(raf).toHaveBeenCalledTimes(2)
  })

  describe('update', () => {
    const pinStart = 10
    const height = 100
    const scrollHeight = 50
    test('should set correct state, if user hasn\'t scrolled beyond pinStart', () => {
      const component = createComponent({pinStart, height, scrollHeight})
      const scrollTo = (scrollTo) => {
        window.pageYOffset = scrollTo
        component.instance().update()
      }
      scrollTo(0)
      scrollTo(pinStart / 2)
      expect(component.state()).toEqual({transform: 0, stickyTop: height - scrollHeight})
    })

    test('should set correct state, if user has scrolled down to pinStart + scrollHeight/2', () => {
      const component = createComponent({pinStart, height, scrollHeight})
      const scrollTo = (scrollTo) => {
        window.pageYOffset = scrollTo
        component.instance().update()
      }

      scrollTo(0)
      scrollTo(pinStart + scrollHeight / 2)
      expect(component.state()).toEqual({transform: -scrollHeight / 2, stickyTop: height - scrollHeight})
    })

    test('should set correct state, if user has scrolled down and back up again', () => {
      const component = createComponent({pinStart, height, scrollHeight})
      const scrollTo = (scrollTo) => {
        window.pageYOffset = scrollTo
        component.instance().update()
      }

      scrollTo(pinStart)
      expect(component.state()).toEqual({transform: 0, stickyTop: height - scrollHeight})

      const offset = 5
      scrollTo(pinStart + scrollHeight)
      // Header is completely transformed to the top
      expect(component.state()).toEqual({transform: -scrollHeight, stickyTop: height - scrollHeight})

      scrollTo(pinStart + scrollHeight - offset)
      // Header is partially transformed
      expect(component.state()).toEqual({
        transform: -(scrollHeight - offset),
        stickyTop: height - scrollHeight + offset
      })

      scrollTo(pinStart)
      expect(component.state()).toEqual({transform: 0, stickyTop: height - scrollHeight})
    })
  })
})
