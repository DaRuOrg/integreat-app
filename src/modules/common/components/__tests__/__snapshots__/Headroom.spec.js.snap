// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Headroom should render with no stickyAncestor supplied 1`] = `
<React.Fragment>
  <div
    className="headroom"
    style={
      Object {
        "transform": "translateY(0px)",
      }
    }
  >
    <MockNode />
  </div>
</React.Fragment>
`;

exports[`Headroom should render with values from its state 1`] = `
<Headroom
  height={100}
  pinStart={0}
  scrollHeight={50}
  stickyAncestor={<MockAncestor />}
>
  <div
    className="headroom"
    style={
      Object {
        "transform": "translateY(42px)",
      }
    }
  >
    <MockNode>
      <div />
    </MockNode>
  </div>
  <MockAncestor
    stickyTop={24}
  >
    <div />
  </MockAncestor>
</Headroom>
`;
